lms/static/scripts/frontend_apps/components/dashboard/AssignmentActivity.tsx

```diff
@@ -10,12 +10,46 @@ import { useParams } from 'wouter-preact';
 
 import type { Assignment, StudentsResponse } from '../../api-types';
 import { useConfig } from '../../config';
-import { urlPath, useAPIFetch } from '../../utils/api';
+import { apiCall, urlPath } from '../../utils/api';
 import { formatDateTime } from '../../utils/date';
+import { useFetch } from '../../utils/fetch';
 import { replaceURLParams } from '../../utils/url';
+import type { LoaderOptions } from '../ComponentWithLoaderWrapper';
 import DashboardBreadcrumbs from './DashboardBreadcrumbs';
 import OrderableActivityTable from './OrderableActivityTable';
 
+export function loader({
+  config: { dashboard, api },
+  params: { assignmentId },
+  signal,
+}: LoaderOptions) {
+  const { routes } = dashboard;
+  const { authToken } = api;
+
+  return Promise.all([
+    apiCall<Assignment>({
+      path: replaceURLParams(routes.assignment, {
+        assignment_id: assignmentId,
+      }),
+      authToken,
+      signal,
+    }),
+    apiCall<StudentsResponse>({
+      path: replaceURLParams(routes.assignment_stats, {
+        assignment_id: assignmentId,
+      }),
+      authToken,
+      signal,
+    }),
+  ]).then(([assignment, students]) => ({ assignment, students }));
```

Comment:
For now, I made all loaders return a single promise with all the data to be displayed, which needs to be fully awaited before getting the component rendered.

However, we should split this, and make it somehow return the one promise with the "main" data that needs to be awaited, and another promise (or set of promises?) with the less important data which we don't need to wait for, and the component itself can handle.

In this particular case, for example, waiting for the students list is important, as it's the main piece of data to display here. However, waiting for the assignment should not block rendering the page, as it will usually load faster, and if it doesn't, we can make this component display it's own small loader until that's ready.

Reply:
This draft PR implements what is described above: https://github.com/hypothesis/lms/pull/6360